### PR TITLE
Add GetClassLoaderProcessor (SonarSource rule 3032)

### PR DESCRIPTION
**Pull request by henry-lp**
Add a processor for [Rule 3032](https://rules.sonarsource.com/java/RSPEC-3032)

this processor will only transform `.java` files which directly import anything from  `javax` packages. 

Transformation:
```diff
+    ClassLoader d = Thread.currentThread().getContextClassLoader();
-    ClassLoader d = this.getClass().getClassLoader(); // Noncompliant
+    public void case1() {
+        ClassLoader c = Thread.currentThread().getContextClassLoader();
+
+    }
-    public void case1() {
-        ClassLoader c = this.getClass().getClassLoader(); // Noncompliant
-    }
+    public void case2() throws ClassNotFoundException {
+        Thread.currentThread().getContextClassLoader().loadClass("anotherclass");
+    }
-    public void case2() throws ClassNotFoundException{
-        Dummy.class.getClassLoader().loadClass("anotherclass"); // Noncompliant
-    }
+    abstract class InnerClass {
+        ClassLoader f = Thread.currentThread().getContextClassLoader();
+        public void hello() {
+            Thread.currentThread().getContextClassLoader();
+        }
+    }
-    abstract class InnerClass {
-        ClassLoader f = this.getClass().getClassLoader(); // Noncompliant 
-        public void hello() {
-            this.getClass().getClassLoader(); // Noncompliant
-       }
-    }
```
Minor known issue: it seems Spoon import `javax.ejb.embeddable.EJBContainer` again after transformation.

### CODE REVIEW

File: docs/HANDLED_RULES.md [hunk 1]
```diff
@@ -141,6 +141,42 @@ Example:
 
 -----
 
+#### JEE applications should not "getClassLoader" ([Sonar Rule 3032](https://rules.sonarsource.com/java/RSPEC-3032))
+
+Using the standard getClassLoader() may not return the right class loader in Java Enterprise Edition context, i.e. the code use any of these [javax packages](https://docs.oracle.com/javaee/6/api/overview-summary.html). Instead, go through the currentThread. In particular, this processor will transform the code in `.java` files if they appear to use the `javax` package.
```
Comment (fermadeiral) [inline docs/HANDLED_RULES.md:146]: Format getClassLoader() like `getClassLoader()` (and all source code elements for further PRs).

Comment (fermadeiral) [inline docs/HANDLED_RULES.md:146]: "i.e. the code use any of these [javax packages](https://docs.oracle.com/javaee/6/api/overview-summary.html)" -> the connection of this sentence with the previous one is strange, there is a gap. Try to be clearer.

"these [javax packages]" -> "[these javax packages]"

"Instead, go through the currentThread." -> this sentence is not good. It seems you are asking our users to do something. Replace this sentence by "The calls to `getClassLoader()`, such as `this.getClass().getClassLoader()` and `Dummy.class.getClassLoader()`, are replaced by `Thread.currentThread().getContextClassLoader()`.

"In particular, this processor will transform the code in `.java` files if they appear to use the `javax` package." -> why if they "appear"?

General note for writing this description here: try to be succinct, and to say what exactly is changed by the processor. No much details on technical aspects are needed.

File: docs/HANDLED_RULES.md [hunk 2]
```diff
@@ -141,6 +141,42 @@ Example:
...
++        public void hello() {
++            Thread.currentThread().getContextClassLoader();
++        }
++    }
+-    abstract class InnerClass {
+-        ClassLoader f = this.getClass().getClassLoader(); // Noncompliant 
+-        public void hello() {
+-            this.getClass().getClassLoader(); // Noncompliant
+-       }
+-    }
```
Comment (fermadeiral) [inline docs/HANDLED_RULES.md:176]: Too many examples. Here, it's just to give one or two examples of transformations. Select the best cases. Here, I think that the best are:

-     ClassLoader d = this.getClass().getClassLoader(); // Noncompliant
+    ClassLoader d = Thread.currentThread().getContextClassLoader();

-     Dummy.class.getClassLoader().loadClass("anotherclass"); // Noncompliant
+    Thread.currentThread().getContextClassLoader().loadClass("anotherclass");

Also, at least here in the documentation, always show the deleted line and then the added line, and not the other way around as you did. The deleted line followed by the added line represents better the "before and after".

File: src/main/java/sonarquberepair/processor/spoonbased/GetClassLoaderProcessor.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package sonarquberepair.processor.spoonbased;
+
+import sonarquberepair.processor.SQRAbstractProcessor;
+
+import spoon.processing.AbstractProcessor;
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/processor/spoonbased/GetClassLoaderProcessor.java:5]: Unused import.

File: src/test/resources/GetClassLoader.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+		ClassLoader f = this.getClass().getClassLoader(); // Noncompliant
+
+		public void hello() {
+			this.getClass().getClassLoader(); // Noncompliant
+		}
+	}
+
+	public void usingEJB() {
+		EJBContainer container = new EJBContainer();
+	}
```
Comment (fermadeiral) [inline src/test/resources/GetClassLoader.java:25]: This last method seems irrelevant. What is it?

Comment (henry-lp) [reply src/test/resources/GetClassLoader.java:25]: it shows that the code is in JEE context as it uses an item from `javax` :) . If this piece does not exist then the import will not be parsed in Spoon later as it's regarded as unused import.

File: docs/HANDLED_RULES.md (hunk 1 above)
Comment (henry-lp) [reply docs/HANDLED_RULES.md:146]: Thanks for the input :) . Fixed.

File: README.md [hunk 5]
```diff
@@ -4,7 +4,7 @@ Sonarqube-repair is a collection of java code analyses and transformations made
 
 ## Handled rules
 
-Sonarqube-repair can currently repair violations of 8 rules of which 6 are labeled as `BUG` and 2 as `Code Smell`. [Check out the handled rules](/docs/HANDLED_RULES.md).
+Sonarqube-repair can currently repair violations of 9 rules of which 7 are labeled as `BUG` and 2 as `Code Smell`. [Check out the handled rules](/docs/HANDLED_RULES.md).
```
Comment (fermadeiral) [inline README.md:7]: This same change should be also done in `docs/HANDLED_RULES.md`.

File: docs/HANDLED_RULES.md [hunk 6]
```diff
@@ -141,6 +141,19 @@ Example:
 
 -----
 
+#### JEE applications should not "getClassLoader" ([Sonar Rule 3032](https://rules.sonarsource.com/java/RSPEC-3032))
+
+Using the standard `getClassLoader()` may not return the right class loader in Java Enterprise Edition context. Instead, `getClassLoader()` usage such as `this.getClass().getClassLoader()` and `Dummy.class.getClassLoader()` should be replaced by `Thread.currentThread().getClassLoader()`. In particular, such replacement only occurs if the `.java` files use the `javax` package in its imports.
```
Comment (fermadeiral) [inline docs/HANDLED_RULES.md:146]: "Thread.currentThread().getClassLoader()" -> "Thread.currentThread().getContextClassLoader()"

"the `.java` files use" -> "the `.java` file uses"

File: src/test/resources/GetClassLoader.java (hunk 4 above)
Comment (fermadeiral) [reply src/test/resources/GetClassLoader.java:25]: Understood.

**Comment by henry-lp** [comment]
done :) thanks for the great review.

**Comment by fermadeiral** [comment]
> done :) thanks for the great review.

Henry, there are 3 additional comments.

**Comment by henry-lp** [comment]
sorry I forgot to push

File: docs/HANDLED_RULES.md (hunk 6 above)
Comment (fermadeiral) [reply docs/HANDLED_RULES.md]: Henry, you didn't fix everything here. See that you say "its imports" thus you need to replace "the .java files use" to "the .java file uses".

Comment (henry-lp) [reply docs/HANDLED_RULES.md]: done :) .

**Comment by fermadeiral** [comment]
@henry-lp, congratulations on your first processor!

**Comment by henry-lp** [comment]
thanks a lot, Fernanda :) .
